Added support for AWS CloudFront in Frontend cache invalidation module

robmoorman opened the pull request:
Cache invalidation support for CloudFront, the Content Delivery Network of Amazon Web Services.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/torchbox/wagtail/1845)
<!-- Reviewable:end -->

kaedroho commented on setup.py:
```diff
@@ -21,6 +21,7 @@
 
```
Why the mode change for this file?

prydie commented on wagtail/contrib/wagtailfrontendcache/backends.py:
```diff
@@ -79,3 +82,27 @@ def purge(self, url):
+            )
+        except botocore.exceptions.ClientError as e:
+            logger.error("Couldn't purge '%s' from Cloudfront. ClientError: %s %s", url, e.response['Error']['Code'], e.response['Error']['Message'])
+            return
```
Method will return anyway here without explicit `return` statement.

kaedroho commented:
Thanks! This is really nice work. Sorry it's taken me a while to get round to testing it.

After some testing, it appears that CloudFront doesn't take full URLs, we can only send it the path component of the URL.

``purge_page_from_cache()`` passes in the full URL to ``purge()`` so it doesn't work at the moment:

```
>>> purge_page_from_cache(p)
Couldn't purge 'http://<snip>.cloudfront.net/test/' from Cloudfront. ClientError: InvalidArgument Your request contains one or more invalid invalidation paths.
```

However, purging ``/test/`` using ``purge_url_from_cache`` works fine:
```
>>> purge_page_from_cache('/test/')
<no output>
```

We could, of course, just snip off the hostname and scheme from the URL, but I think we could do better to support multi-site installations.

How about allowing users to configure a mapping of hostname => distribution ids in their settings? (but still support the current single distribution id approach for single-site installations)

For example:
```python
    WAGTAILFRONTENDCACHE = {
        'cloudfront': {
            'BACKEND': 'wagtail.contrib.wagtailfrontendcache.backends.CloudfrontBackend',
            'DISTRIBUTION_IDS': {
                'www.mysite.com': 'distribution-id',
                'www.othermysite.com': 'other-distribution-id',
            }
        },
    }
```

robmoorman commented:
Thanks for your review and proper testing.
I like the idea of distribution id's per site, however it probably clashes with the ability where authorized users can change the site domain at runtime. Ideally you would have an extra input for it near the domain name within the site settings in Wagtail. However, this required an extendable settings model when you use CloudFront caching... probably also not the best solution.
In that way I see your suggestions is the best fit. Like to hear what your opinion is about this and maybe you have another suggestions to make this distribution id content manageable.
In the meantime I will fix the path component.

robmoorman commented:
Fixed the mapping of hostnames and the path element.
Added some tests to make sure the right distributionId is called with a path (instead of full url).
Also added a note to make sure users should include www in their mapping.

tomdyson commented:
Thanks, @robmoorman! @kaedroho, can you look at this again please?

kaedroho commented on tox.ini:
```diff
@@ -19,6 +19,7 @@ basepython =
     py35: python3.5
 
 deps =
+	boto3==1.1.4
```
Should be 4 spaces here instead of tab

kaedroho commented on wagtail/contrib/wagtailfrontendcache/backends.py:
```diff
@@ -83,3 +87,50 @@ def purge(self, url):
+        url_parsed = urlparse(url)
+        distribution_id = None
+
+        if type(self.cloudfront_distribution_id) is dict:
```
This should be ``if isinstance(self.cloudfront_distribution_id, dict)``

kaedroho commented on wagtail/contrib/wagtailfrontendcache/backends.py:
```diff
@@ -83,3 +87,50 @@ def purge(self, url):
+            if host in self.cloudfront_distribution_id:
+                distribution_id = self.cloudfront_distribution_id.get(host)
+            else:
+                raise ImproperlyConfigured("hostname %s" % host)
```
Looks like this line was accidentally committed

kaedroho commented on wagtail/contrib/wagtailfrontendcache/tests.py:
```diff
@@ -43,6 +46,41 @@ def test_cloudflare(self):
+        backends.get('cloudfront').purge('http://www.wagtail.io/home/events/christmas/')
+
+        _create_invalidation.assert_any_call('frontend', '/home/events/christmas/')
+
```
Could we have a test case for when a cache invalidation is performed on a domain that isn't in the ``DISTRIBUTION_ID`` mapping?

robmoorman commented:
Thanks @kaedroho , I'll pick up your feedback on this PR this week!

kaedroho commented:
Reviewed 2 of 7 files at r1, 1 of 4 files at r4.
Review status: 3 of 7 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/torchbox/wagtail/1845)*
<!-- Sent from Reviewable.io -->

robmoorman commented:
he @kaedroho seems like some checks are failing caused by other parts of code from the master branch

gasman commented:
@robmoorman Have fixed the isort errors on master now, so a rebase should hopefully fix it here too.

kaedroho commented:
Reviewed 1 of 7 files at r1, 1 of 4 files at r4, 1 of 2 files at r5.
Review status: 6 of 7 files reviewed at latest revision, 1 unresolved discussion.

---

*[wagtail/contrib/wagtailfrontendcache/backends.py, line 113 \[r6\]](https://reviewable.io:443/reviews/torchbox/wagtail/1845#-KIgVIanihuLS3rEK4Ps:-KIgVIaoxb8mseV1QJWN:-882815873) ([raw file](https://github.com/torchbox/wagtail/blob/db2b9191b8ca10039acd1c75661c0aac54108937/wagtail/contrib/wagtailfrontendcache/backends.py#L113)):*
> ```Python
>                 distribution_id = self.cloudfront_distribution_id.get(host)
>             else:
>                 logger.error("Couldn't purge '%s', hostname '%s' not found in mapping", url, host)
> ```

We should point out in this message that it's cloudfront that's failing and that it's a problem with DISTRIBUTION_ID

How about: Couldn't purge '%s' from CloudFront. Hostname '%s' not found in ``DISTRIBUTION_ID`` mapping

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/torchbox/wagtail/1845)*
<!-- Sent from Reviewable.io -->

kaedroho commented on wagtail/contrib/wagtailfrontendcache/backends.py:
```diff
@@ -84,3 +88,49 @@ def purge(self, url):
+            if host in self.cloudfront_distribution_id:
+                distribution_id = self.cloudfront_distribution_id.get(host)
+            else:
+                logger.error("Couldn't purge '%s', hostname '%s' not found in mapping", url, host)
```
We should point out in this message that it's cloudfront that's failing and that it's a problem with DISTRIBUTION_ID

How about: "Couldn't purge '%s' from CloudFront. Hostname '%s' not found in DISTRIBUTION_ID mapping"?

robmoorman replied:
Good point, I'll keep it consistent with the other error messages (like CloudFlare)

kaedroho commented on wagtail/contrib/wagtailfrontendcache/backends.py:
```diff
@@ -84,3 +88,52 @@ def purge(self, url):
+            if host in self.cloudfront_distribution_id:
+                distribution_id = self.cloudfront_distribution_id.get(host)
+            else:
+                logger.error(
```
It's possible that someone might want cloudfront on one site but not another, so perhaps this should be info rather than error?

kaedroho commented on wagtail/contrib/wagtailfrontendcache/backends.py:
```diff
@@ -84,3 +88,52 @@ def purge(self, url):
+                distribution_id = self.cloudfront_distribution_id.get(host)
+            else:
+                logger.error(
+                    "Couldn't purge '%s' from CloudFront, Hostname '%s' not found in the DISTRIBUTION_ID mapping",
```
Minor inconsistency, the comma after CloudFront should be a full stop.

kaedroho commented on setup.py:
```diff
@@ -21,6 +21,7 @@
 
 
 install_requires = [
+    "boto3>=1.1,<1.2",
```
Sorry for not pointing this out earlier, I think we should avoid adding ``boto3`` as a hard dependency as I expect most people won't be using cloudfront.

kaedroho commented:
Tested and works well!

Spotted a few more things, but I think this will be ready to merge once they are resolved.

Thanks!

kaedroho commented:
Made some tweaks here: https://github.com/torchbox/wagtail/compare/torchbox:master...kaedroho:feature/cloudfront-tweaks

@robmoorman could you review?
